assert: improve assertion errors

BridgeAR opened the pull request:
Please have a look at the commit messages for details.

This makes error messages in case of an error are improved while using `assert.throws()` and `assert.rejects()`. It also improves the error message for simple assertions when used in combination with `new Function` and adds a deprecation warning for validation functions that return anything besides `true`.

This should IMO be semver-minor, since it'll only impact code that fails the assertion (e.g., while writing new code). No passing assertion call is impacted by this.

@nodejs/assert PTAL

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

targos commented on lib/assert.js:
```diff
@@ -549,30 +549,38 @@ function compareExceptionKey(actual, expected, key, message, keys, fn) {
+        actual,
+        expected,
+        message: message || `The regular expression ${inspect(expected)} did ` +
+                            `not match the input:\n\n${inspect(str)}\n`,
```
I'm not sure but I think it should be the other way: The input did not match the regular expression.
As an argument, the `match` method is on `String`s prototype.

BridgeAR replied:
I agree and I tried to find an appropriate wording for it as well but I ended up with this because I could not find a good brief description that allowed to name the RegExp first. Do you have a suggestion to reword this?

BridgeAR replied:
The reason for preferring the RegExp first is that it is mostly short and on a single line. The input could be a long multi line string.

BridgeAR replied:
@targos what do you think about: `The input did not match the regular expression /foobar/. Input:\n\${input}`?

// cc @Trott

BridgeAR replied:
I just updated the code with my suggestion.

BridgeAR commented:
@nodejs/assert @nodejs/testing PTAL. This is actually quite a decent improvement for some cases.

Trott commented on doc/api/deprecations.md:
```diff
@@ -2467,6 +2467,22 @@ The legacy HTTP parser, used by default in versions of Node.js prior to 12.0.0,
+
+Returning anything besides `true` as result of an `assert.throws()` or
+`assert.rejects()` validation function is deprecated. Doing so results in the
+caught error from being thrown again. That is likely not desirable. Please throw
```
This presumably is either:

"Doing so prevents the caught error from being thrown again."

Or:

"Doing so results in the caught error being thrown again."

Whichever one is correct, can you update the wording as appropriate?

Trott commented on doc/api/assert.md:
```diff
@@ -1210,10 +1210,15 @@ assert.throws(
+    assert(/value/.test(err));
+    // Note: returning anything from validation functions besides `true` is
+    // deprecated.
+    // Doing so results in the caught error from being thrown again. That is
```
```console
    // Doing so results in the caught error being thrown again. That is
```

Trott commented:
Introduction of any runtime deprecation is semver major and I don't know that this should be an exception?

jasnell commented:
I agree with @Trott on semver-major. We should be consistent with the policy

Trott commented:
I'm not sure I agree with this. If the validation function returns `false`, then the error should be re-thrown (or an AssertionError thrown). I don't think we want to change that. I think that's the right result. Validation functions should return `true` or `false` (and we should *probably* accept truthy/falsy values too, but I'm less opinionated about that). If the validation function returns `false`, then the behavior I would expect is an AssertionError to be thrown or the error would be re-thrown. Anything else is surprising and non-ergonomic in my opinion.

It's more than possible that I either haven't thought about this with sufficient care or else am simply missing the way this change improves things for the average user. But the idea that a validation function shouldn't return `false` is surprising and not very ergonomic IMO. The benefit of enforcing that needs to be enormous.

cjihrig commented:
https://github.com/nodejs/node/pull/27781/commits/a584c57c3733636db1b5b81c045f2d98a5fc4647 seems like an unrelated change that shouldn't be part of the larger PR, and I disagree with it in general because I think it decreases visibility into our tests.

I also agree with @trott's https://github.com/nodejs/node/pull/27781#issuecomment-497094819. I don't think that's a change we should make.

BridgeAR commented:
I just updated the PR and incorporated the feedback. I removed the deprecation completely and I'll open a separate PR that changes the behavior directly to an actually more desirable one (throwing an `AssertionError` that the user returned other than `true` to give a better context) instead of having a deprecation in this case.

@cjihrig I also removed the commit that you mentioned. I just added it, since without it we receive a cryptic error message. I guess you would prefer to just throw a more meaningful error in that case instead of just logging it?

@nodejs/assert PTAL

Trott commented on doc/api/assert.md:
```diff
@@ -1210,10 +1210,14 @@ assert.throws(
+    assert(/value/.test(err));
+    // Returning anything from validation functions besides `true` is not
+    // recommended. Doing so results in the caught error being thrown again.
+    // That is mostly not the desired outcome. Throw an error about the specific
```
```suggestion
    // That is usually not the desired outcome. Throw an error about the specific
```

Trott commented:
Not opposed to this, but I'm still not sure that re-throwing the error is all that undesirable? Pondering some more...

BridgeAR commented:
@Trott right now the user loses pretty much all context why this error occurred and there is no reference to any assertion going on at all. This struck me multiple times in tests and it takes more time to figure out what actually happened. By throwing inside of the validation function in case a specific validation fails, that's not a problem at all. That's why I definitely believe we should recommend this.

BridgeAR commented:
Resumed build: https://ci.nodejs.org/job/node-test-commit/29385/

Trott commented:
Landed in 8149656...0f900405e52061136f7b5d69e6f2453309064c8a

BethGriggs commented:
@BridgeAR, should this land on v10.x? Please add the `lts-watch` label if so
